WebKit/Source/Platform/IPC should be testable with C++ unit tests

kkinnunen-apple opened the pull request:
#### e251ba5a08e4c3a27f689c6db01126cf5af2078f
<pre>
WebKit/Source/Platform/IPC should be testable with C++ unit tests
<a href="https://bugs.webkit.org/show_bug.cgi?id=243780">https://bugs.webkit.org/show_bug.cgi?id=243780</a>
rdar://problem/98717673

Reviewed by Elliott Williams and Alexey Proskuryakov.

Add TestIPC executable which links to WebCore.framework and
libWebKitPlatform.a.

Add an example test for IPC code.

* Source/WebCore/Configurations/WebCore.xcconfig:
* Tools/Scripts/run-api-tests:
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort.path_to_api_test_binaries):
* Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig: Added.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionBufferTests.cpp: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/mac/mainMac.mm:
(main):

Canonical link: <a href="https://commits.webkit.org/254030@main">https://commits.webkit.org/254030@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4466fba57c5c1db24bbf2cbcef4dfec84c42ae41

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87698 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31783 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/18438 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/96894 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/150702 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/91667 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30141 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/26251 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79797 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/91641 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93317 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/24347 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/74448 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/24347 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/5/builds/91302 "Passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/79310 "Passed tests") | [❌ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/79503 "Found 1 new API test failure: TestWebKitAPI.VideoControlsManager.VideoControlsManagerMultipleVideosScrollPlayingVideoWithSoundOutOfView") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/67196 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27836 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/13324 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27802 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/14339 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2833 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29497 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/37232 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29397 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33614 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig:
```diff
@@ -0,0 +1,59 @@
+OTHER_LDFLAGS_PLATFORM_ = -framework Cocoa -framework Carbon -framework CoreServices;
+
+// FIXME: This should not be built on iOS. Instead we should create and use a TestWTF application.
+OTHER_LDFLAGS_PLATFORM_cocoatouch = -framework CoreGraphics -framework MobileCoreServices -framework UIKit ;
```
Extra space before the semicolon here.

aproskuryakov commented on Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig:
```diff
@@ -0,0 +1,59 @@
+WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS_iphonesimulator = $(WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS$(WK_IOS_14));
+WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS_maccatalyst = $(WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS$(WK_IOS_14));
+WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS_IOS_SINCE_14 = -framework UniformTypeIdentifiers;
+WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS_macosx = $(WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS$(WK_MACOS_1100));
```
I doubt that `WK_UNIFORM_TYPE_IDENTIFIERS` is relevant in this project, and I'm quite certain that iOS 14 and pre-macOS 11 isn't.

aproskuryakov commented on Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig:
```diff
@@ -0,0 +1,59 @@
+OTHER_LDFLAGS = $(inherited) $(WK_UIKITMACHELPER_LDFLAGS) $(OTHER_LDFLAGS_PLATFORM_$(WK_COCOA_TOUCH)) $(WK_UNIFORM_TYPE_IDENTIFIERS_LDFLAGS);
+OTHER_LDFLAGS_PLATFORM_ = -framework Cocoa -framework Carbon -framework CoreServices;
+
+// FIXME: This should not be built on iOS. Instead we should create and use a TestWTF application.
```
TestWTF?

aproskuryakov reviewed (approved):
LGTM, but let's have Elliott weigh in on whether project dependencies are right for workspace (i.e. normal) builds.

emw-apple commented on Tools/Scripts/webkitpy/port/darwin.py:
```diff
@@ -247,6 +247,9 @@ def _path_to_image_diff(self):
 
+    def path_to_api_test_binaries(self):
+        return {binary: self._build_path(binary) for binary in super(DarwinPort, self).API_TEST_BINARY_NAMES + ['TestIPC']}
+
```
nit: For legibility, I would prefer just overriding API_TEST_BINARY_NAMES instead of inheriting like this. We already do it for other ports.
```suggestion
    API_TEST_BINARY_NAMES = ['TestWTF', 'TestWebKitAPI', 'TestIPC']

```

emw-apple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -5418,6 +5523,7 @@
 				8DD76F9B0486AA7600D96B5E /* Frameworks */,
 				8DD76F9E0486AA7600D96B5E /* CopyFiles */,
 				BCB9F4FB112384C000A137E0 /* Copy Resources */,
+				7B9FC5CA28A52CCD007570E7 /* Embed Frameworks */,
```
Does TestWebKitAPI really need an "Embed Frameworks" phase? It's not a bundle with a frameworks directory, and it would normally be executed alongside WebCore.framework and others.

emw-apple commented on Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
```diff
@@ -5357,6 +5444,24 @@
+			buildPhases = (
+				7B9FC39B28A26137007570E7 /* Sources */,
+				7B9FC3A328A26137007570E7 /* Frameworks */,
+				7B9FC5B428A52224007570E7 /* Embed Frameworks */,
```
Does TestIPC really need an "Embed Frameworks" phase? It's not a bundle with a frameworks directory, and it would normally be executed alongside WebCore.framework and others.

emw-apple reviewed (approved):
LGTM once the "Embed Frameworks" phases are fixed.

kkinnunen-apple committed:
WebKit/Source/Platform/IPC should be testable with C++ unit tests
https://bugs.webkit.org/show_bug.cgi?id=243780
rdar://problem/98717673

Reviewed by Elliott Williams and Alexey Proskuryakov.

Add TestIPC executable which links to WebCore.framework and
libWebKitPlatform.a.

Add an example test for IPC code.

* Source/WebCore/Configurations/WebCore.xcconfig:
* Tools/Scripts/run-api-tests:
* Tools/Scripts/webkitpy/port/darwin.py:
(DarwinPort.path_to_api_test_binaries):
* Tools/TestWebKitAPI/Configurations/TestIPC.xcconfig: Added.
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/StreamConnectionBufferTests.cpp: Added.
(TestWebKitAPI::TEST):
* Tools/TestWebKitAPI/mac/mainMac.mm:
(main):

Canonical link: https://commits.webkit.org/254030@main
